posthog/models/insight_caching_state.py

```diff
@@ -0,0 +1,38 @@
+from django.db import models
+
+from posthog.models.team import Team
+from posthog.models.utils import UUIDModel
+
+
+class InsightCachingState(UUIDModel):
+    class Meta:
+        indexes = [models.Index(fields=["cache_key"], name="filter_by_cache_key_idx")]
+        constraints = [
+            models.UniqueConstraint(
+                fields=["insight"],
+                name=f"unique_insight_for_caching_state_idx",
+                condition=models.Q(("insight__isnull", False)),
+            ),
+            models.UniqueConstraint(
+                fields=["insight", "dashboard_tile"],
+                name=f"unique_dashboard_tile_idx",
+                condition=models.Q(("dashboard_tile__isnull", False)),
+            ),
+        ]
+
+    team: models.ForeignKey = models.ForeignKey(Team, on_delete=models.CASCADE)
+
+    insight = models.ForeignKey("posthog.Insight", on_delete=models.CASCADE, related_name="caching_state", null=False)
```

Comment:
Should this be one to many instead, so there's a single unique "InsightCachingState" with a single cache_key, that all relevant insights linked to?

I guess you're trying to solve the issue of the same insight having different refresh ages on dashboards, so maybe you're approaching it from the other way?

Reply:
I think it's better to go 1-to-1 instead and deduplicate calculations during scheduling of work. We had this discussion with Paul already, though not sure he's convinced.

Reasoning: I think keeping this model up-to-date will be a pain if we start creating "through" tables or store it in arrays. With the proposed schema it's easy to verify whether everything that should have an entry has one and what state it is in.

Reply:
> We had this discussion with Paul already, though not sure he's convinced.

It's cheap to update or find a set of items with the same cache key all at once. 

Even though in theory we operate on the cache key alone when dealing with caching. The code is always aware of which insight or tile it is dealing with. 

* We discover what needs to be operated on by the cache starting from tiles/insights
* We discover whether something is cached or how old its cache is starting from tiles/insights
* We update the cache key on tiles/insights when they're saved in case their filters changed so we can get to them from their cache key - now I guess instead we'd upsert the cache state for the new cache key
* We correct the cache key on the insight/tile if its out of synch with the calculated cache when updating the cache. This has always felt like a sticking plaster on a hard to identify bug though
* We update the cache state of the set of insights and tiles that match a key when processing the cache. Now we would update the set of matching cache states
* I'm assuming we only upsert cache states in the cache processing and not during insight editing/exploration so we don't end up with dangling cache states. We'd have the cache key/results pair in the cache but wouldn't rack it for periodic updates

It didn't match my initial intuition but if it means we can eventually stop storing cachekey and other state on the insight/tile then I'm 100% on board.